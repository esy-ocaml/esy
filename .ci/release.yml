steps:
  - task: DownloadBuildArtifacts@0
    displayName: 'Download Linux Artifacts'
    inputs:
      artifactName: Linux
      downloadPath: '_release'

  - script: 'mkdir _release/platform-linux'
    displayName: 'Create _release/platform-linux'

  - script: 'tar -xzf _release/Linux/*.tgz -C _release/platform-linux'
    displayName: 'untar Linux'

  - task: DownloadBuildArtifacts@0
    displayName: 'Download macOS Artifacts'
    inputs:
      artifactName: macOS
      downloadPath: '_release'

  - script: 'mkdir _release/platform-darwin'
    displayName: 'Create _release/platform-darwin'

  - script: 'tar -xzf _release/macOS/*.tgz -C _release/platform-darwin'
    displayName: 'untar macOS'

  - task: DownloadBuildArtifacts@0
    displayName: 'Download Windows Artifacts'
    inputs:
      artifactName: Windows
      downloadPath: '_release'
    continueOnError: true

  - script: 'mkdir _release/platform-windows-x64'
    displayName: 'Create _release/platform-windows-x64'
    continueOnError: true

  - script: 'tar -xzf _release/Windows/*.tgz -C _release/platform-windows-x64'
    displayName: 'untar Windows'
    continueOnError: true

  - script: 'node scripts/pipelines-release.js'
    displayName: 'node scripts/pipelines-release.js'

  - script: 'npm pack'
    displayName: 'npm pack'
    workingDirectory: '_release'

  - task: PublishBuildArtifacts@1
    displayName: 'Publish packaged release'
    inputs:
      PathtoPublish: '_release/esy-*.tgz'
      ArtifactName: release
