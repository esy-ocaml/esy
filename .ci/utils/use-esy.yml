# steps to install esy globally

steps:
- script: "yarn global add @esy-nightly/esy@0.6.8-bbdfb8"
  displayName: "Install esy (via yarn)"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')
  
- script: |
    yarn config set prefix C:\
    yarn global add @esy-nightly/esy@0.6.8-bbdfb8
  displayName: "Install esy (via yarn)"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')

- bash: |
    YARN_BIN_DIR=$(yarn global bin)
    echo $YARN_BIN_DIR
    echo "##vso[task.setvariable variable=YARN_BIN_DIR]$YARN_BIN_DIR"
  displayName: "Setting $YARN_BIN_DIR"

- bash: |
    echo "##vso[task.setvariable variable=PATH]$PATH:$YARN_BIN_DIR" 
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')

- powershell: Write-Host "##vso[task.setvariable variable=PATH;]${env:PATH};${env:YARN_BIN_DIR}"
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')
