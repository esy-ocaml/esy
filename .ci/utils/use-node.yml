# steps to use node on agent

steps:
  - task: NodeTool@0
    displayName: "Use Node 15.x"
    inputs:
      versionSpec: 15.x
