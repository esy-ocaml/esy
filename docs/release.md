---
id: release
title: Building Releases
---

