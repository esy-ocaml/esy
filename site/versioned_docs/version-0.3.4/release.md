---
id: version-0.3.4-release
title: Building Releases
original_id: release
---

Applications developed with esy can have complex dependencies at runtime; not
just code linked statically but also executables, dynamically loaded libraries
and other resources such as images, translations and so on.

This requires a special release process which bundles all those dependencies
together in an end user installable package.

There's `esy release` command which produces an npm package ready to be
published which:

1. Packages project's and all dependencies' pre built artifacts.
2. Exposes a specified set of project's executables as commands available to end
   users.

The very important property of such packages is that only a configured set of
executable is being exposed to an end user, all other executables and libraries
are "private" to the release. That means multiple releases can be installed on a
user's machine with different versions of the same software without interfering
between each other.

## Producing Releases

To configure a release one should add `esy.release.releasedBinaries`
configuration field to a `package.json`:

```json
"esy": {
  "release": {
    "releasedBinaries": [
      "refmt"
    ]
  }
}
```

Such field lists all executable names which must be made available on `$PATH` when
release is installed on a user's machine.

After that configuration is done the only command to run is:

```bash
% esy release
```

Which produces a `_release` directory with a ready to be published npm package
with pre built binaries for the current platform.

> Currently release can only contain an application built for a single platform
> (macOS, Linux or Windows). This restriction will be lifted in the future.

To publish such package to npm:

```bash
% (cd _release && npm publish)
```

## Excluding Dependencies

Not all dependencies of an esy project are needed at runtime. Therefore it makes
sense to exclude them from the release which will make release size smaller.

> We are working on a way to make esy configuration more granular which would
> allow to automatically exclude non-runtime dependencies from releases. We are
> not there yet.

There's `esy.release.deleteFromBinaryRelease` configuration field in
`package.json` which allows to specify a list of glob patterns against artifact
names to be excluded from a release package:

```json
"esy": {
  "release": {
    "releasedBinaries": [ ... ]
    "deleteFromBinaryRelease": [
      "ocaml-*",
      "*jbuilder-*"
    ]
  }
}
```

The configuration above will make sure OCaml toolchain, jbuilder package won't
be present in a release.
