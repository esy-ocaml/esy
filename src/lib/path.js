/**
 * @flow
 */

import invariant from 'invariant';
const path = require('path');

/**
 * Absolute path.
 */
export opaque type AbsolutePath: ConcretePath = string;

/**
 * Path with no `.` or `..` or any bash variables in it.
 */
export opaque type ConcretePath: Path = string;

/**
 * Path which can contain bash/make variables in it, cannot be used in
 * filesystem operations.
 */
export opaque type AbstractPath: Path = string;

export opaque type Path: string = AbsolutePath | ConcretePath | AbstractPath;

export function absolute(p: string): AbsolutePath {
  p = sanitizeAbsolutePath(p);
  return p;
}

export function concrete(p: string): ConcretePath {
  p = sanitizeConcretePath(p);
  return p;
}

export function abstract(p: string): AbstractPath {
  return p;
}

export function join<B: Path | string>(base: B, ...p: string[]): B {
  return (path.join(base, ...p): any);
}

export function basename<P: Path | string>(p: P): string {
  return path.basename(p);
}

export function dirname<P: Path | string>(p: P): P {
  return (path.dirname(p): any);
}

export function length(p: ConcretePath): number {
  return p.length;
}

export function toString(p: Path | string): string {
  return p;
}

export const sep: typeof path.sep = path.sep;

export const relative: typeof path.relative = path.relative;

function sanitizeAbsolutePath(p) {
  p = sanitizeConcretePath(p);
  invariant(path.isAbsolute(p), 'Should be absolute path but got: %s', p);
  return p;
}

function sanitizeConcretePath(p) {
  // NOTE: regexpes are stateful, this is why we are creating them here.
  // Consider resetting their state after matching instead.
  const REMOVE_TRAILING_SLASH_RE = /\/+$/g;
  const ENVIRONMENT_VAR_RE = /\$[a-zA-Z_]/g;
  const DOT_PATH_SEGMENT_RE = /\/\.\//g;
  const DOT_DOT_PATH_SEGMENT_RE = /\/\.\.\//g;
  const SANITIZE_SLASH_RE = /\/+/g;

  invariant(
    DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain "." segments: %s',
    p,
  );
  invariant(
    DOT_DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain ".." segments: %s',
    p,
  );
  invariant(
    ENVIRONMENT_VAR_RE.exec(p) == null,
    'Path should not contain variable references: %s',
    p,
  );
  p = path.normalize(p);
  p = p.replace(REMOVE_TRAILING_SLASH_RE, '');
  p = p === '' ? '/' : p;
  return p;
}
