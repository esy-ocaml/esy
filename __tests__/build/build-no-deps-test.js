/**
 * @flow
 */

import * as path from 'path';
import {initFixtureSync, readDirectory, cleanUp} from '../release/utils';

const fixture = initFixtureSync(path.join(__dirname, 'fixtures', 'no-deps'));

test(`build ${fixture.description}`, async function() {
  const buildStdout = await fixture.esy(['build'], {cwd: fixture.project});
  expect(buildStdout).toMatchSnapshot('build stdout');

  const esyPrefixDir = await readDirectory(path.join(fixture.esyPrefix));
  expect(esyPrefixDir).toMatchSnapshot('esy prefix dir');

  const esyLocalPrefixDir = await readDirectory(fixture.localEsyPrefix);
  expect(esyLocalPrefixDir).toMatchSnapshot('esy local prefix dir');
});

afterAll(cleanUp);
