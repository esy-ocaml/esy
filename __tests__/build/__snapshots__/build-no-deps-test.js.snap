// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build stdout 1`] = `"esy build v0.0.0"`;

exports[`esy local prefix dir 1`] = `
"<root>
| build
| | bin
| | | _build
| | | _install
| | | _shell
| | | build
| | | build-dependencies
| | | build-env
| | | build-exec
| | | command-env
| | | command-exec
| | | fastreplacestring.cpp
| | | fastreplacestring.exe
| | | install
| | | runtime.sh
| | | sandbox-env
| | | sandbox.sb
| | | shell
| store
| | b
| | | no_deps-1.0.0
| | | | _esy
| | | | no-deps
| | i
| | s
| | | no_deps-1.0.0
| | | | bin
| | | | doc
| | | | etc
| | | | lib
| | | | man
| | | | sbin
| | | | share"
`;

exports[`esy prefix dir 1`] = `
"<root>
| 3
| | b
| | i
| | s
| 3_____________________________________________________________________
| | b
| | i
| | s"
`;
